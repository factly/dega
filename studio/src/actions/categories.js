import axios from 'axios';
import {
  ADD_CATEGORIES,
  ADD_CATEGORIES_REQUEST,
  SET_CATEGORIES_LOADING,
  RESET_CATEGORIES,
  CATEGORIES_API,
  GET_CATEGORY,
  UPDATE_CATEGORY,
} from '../constants/categories';
import { addErrorNotification, addSuccessNotification } from './notifications';
import { addMedia } from './media';
import getError from '../utils/getError';

// action to fetch all categories
export const getCategories = (query) => {
  return (dispatch, getState) => {
    const currentSpaceID = getState().spaces?.selected;
    if (currentSpaceID === 0) {
      return;
    }
    dispatch(loadingCategories());
    return axios
      .get(CATEGORIES_API, {
        params: query,
      })
      .then((response) => {
        dispatch(
          addMedia(
            response.data.nodes
              .filter((category) => category.medium)
              .map((category) => category.medium),
          ),
        );
        dispatch(
          addCategoriesList(
            response.data.nodes.map((category) => {
              return { ...category, medium: category.medium?.id };
            }),
          ),
        );
        dispatch(
          addCategoriesRequest({
            data: response.data.nodes.map((item) => item.id),
            query: query,
            total: response.data.total,
          }),
        );
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopCategoriesLoading()));
  };
};

// action to fetch category by id
export const getCategory = (id) => {
  return (dispatch) => {
    dispatch(loadingCategories());
    return axios
      .get(CATEGORIES_API + '/' + id)
      .then((response) => {
        if (response.data.medium) dispatch(addMedia([response.data.medium]));

        dispatch(addCategory(GET_CATEGORY, { ...response.data, medium: response.data.medium?.id }));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopCategoriesLoading()));
  };
};

// action to create category
export const createCategory = (data) => {
  return (dispatch) => {
    dispatch(loadingCategories());
    return axios
      .post(CATEGORIES_API, data)
      .then(() => {
        dispatch(resetCategories());
        dispatch(addSuccessNotification('Category created'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

// action to update category by id
export const updateCategory = (data) => {
  return (dispatch) => {
    dispatch(loadingCategories());
    return axios
      .put(CATEGORIES_API + '/' + data.id, data)
      .then((response) => {
        if (response.data.medium) dispatch(addMedia([response.data.medium]));

        dispatch(
          addCategory(UPDATE_CATEGORY, { ...response.data, medium: response.data.medium?.id }),
        );
        dispatch(addSuccessNotification('Category updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopCategoriesLoading()));
  };
};

// action to delete category by id
export const deleteCategory = (id) => {
  return (dispatch) => {
    dispatch(loadingCategories());
    return axios
      .delete(CATEGORIES_API + '/' + id)
      .then(() => {
        dispatch(resetCategories());
        dispatch(addSuccessNotification('Category deleted'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

export const addCategories = (categories) => {
  return (dispatch) => {
    dispatch(
      addMedia(categories.filter((category) => category.medium).map((category) => category.medium)),
    );
    dispatch(
      addCategoriesList(
        categories.map((category) => {
          return { ...category, medium: category.medium?.id };
        }),
      ),
    );
  };
};

export const loadingCategories = () => ({
  type: SET_CATEGORIES_LOADING,
  payload: true,
});

export const stopCategoriesLoading = () => ({
  type: SET_CATEGORIES_LOADING,
  payload: false,
});

export const addCategory = (type, payload) => ({
  type,
  payload,
});

export const addCategoriesList = (data) => ({
  type: ADD_CATEGORIES,
  payload: data,
});

export const addCategoriesRequest = (data) => ({
  type: ADD_CATEGORIES_REQUEST,
  payload: data,
});

export const resetCategories = () => ({
  type: RESET_CATEGORIES,
});
