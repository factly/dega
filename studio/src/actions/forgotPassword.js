const userId = localStorage.getItem('userId');
export const requestPasswordReset = async (userId) => {
  const response = await fetch(
    `${window.REACT_APP_ZITADEL_AUTHORITY}/v2/users/${userId}/password_reset`,
    {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        Authorization: `Bearer ${window.REACT_APP_ZITADEL_PAT}`,
      },
      body: JSON.stringify({
        sendLink: {
          notificationType: 'NOTIFICATION_TYPE_Email',
        },
      }),
    },
  );

  if (!response.ok) {
    throw new Error('Failed to send reset email');
  }

  return response.json();
};

export const resetPassword = async (userId, newPassword, verificationCode) => {
  const response = await fetch(
    `${window.REACT_APP_ZITADEL_AUTHORITY}/v2/users/${userId}/password`,
    {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        Authorization: `Bearer ${window.REACT_APP_ZITADEL_PAT}`,
      },
      body: JSON.stringify({
        newPassword: {
          password: newPassword,
          changeRequired: false,
        },
        verificationCode: verificationCode,
      }),
    },
  );

  if (!response.ok) {
    throw new Error('Failed to reset password');
  }

  return response.json();
};
