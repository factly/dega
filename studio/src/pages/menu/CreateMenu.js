import React from 'react';
import MenuForm from './components/MenuForm';
import { useDispatch } from 'react-redux';
 
import { createMenu } from '../../actions/menu';
import { Helmet } from 'react-helmet';
import useNavigation from '../../utils/useNavigation';

function CreateMenu() {
  const history = useNavigation();
  const dispatch = useDispatch();
  const onCreate = (values) => {
    dispatch(createMenu(values)).then(() => history('/settings/website/menus'));
  };
  return (
    <div>
      <>
        <Helmet title={'Create Menu'} />
        <MenuForm onCreate={onCreate} />
      </>
    </div>
  );
}

export default CreateMenu;
