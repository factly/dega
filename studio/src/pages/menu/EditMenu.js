import React from 'react';
import MenuEditForm from './components/MenuForm';
import { useDispatch, useSelector } from 'react-redux';
import { Skeleton } from 'antd';
import { updateMenu, getMenu } from '../../actions/menu';

import { useParams } from 'react-router-dom';
import RecordNotFound from '../../components/ErrorsAndImage/RecordNotFound';
import { Helmet } from 'react-helmet';
import useNavigation from '../../utils/useNavigation';

function EditMenu() {
  const history = useNavigation();
  const { id } = useParams();
  const dispatch = useDispatch();

  const { menu, loading } = useSelector((state) => {
    return {
      menu: state.menus.details[id] ? state.menus.details[id] : null,
      loading: state.menus.loading,
    };
  });

  React.useEffect(() => {
    dispatch(getMenu(id));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [id]);

  if (loading) return <Skeleton />;
  if (!menu) {
    return <RecordNotFound />;
  }
  const onUpdate = (values) => {
    dispatch(updateMenu({ ...menu, ...values })).then(() =>
      history(`/settings/website/menus/${id}/edit`),
    );
  };
  return (
    <>
      <Helmet title={`${menu?.name} - Edit Menu`} />
      <MenuEditForm data={menu} onCreate={onUpdate} />
    </>
  );
}

export default EditMenu;
