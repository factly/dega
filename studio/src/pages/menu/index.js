import React from 'react';
import { Space, Button, Row } from 'antd';
import MenuList from './components/MenuList';
import { Link, useLocation } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { getMenus } from '../../actions/menu';
import deepEqual from 'deep-equal';
import getUserPermission from '../../utils/getUserPermission';
import Loader from '../../components/Loader';
import { Helmet } from 'react-helmet';

function Menu() {
  const spaces = useSelector(({ spaces }) => spaces);
  const actions = getUserPermission({ resource: 'menus', action: 'get', spaces });
  const dispatch = useDispatch();
  const query = new URLSearchParams(useLocation().search);
  const [filters, setFilters] = React.useState({
    page: 1,
    limit: 20,
  });
  query.set('page', filters.page);
  window.history.replaceState({}, '', `${window.PUBLIC_URL}${useLocation().pathname}?${query}`);
  const { menus, total, loading } = useSelector((state) => {
    const node = state.menus.req.find((item) => {
      return deepEqual(item.query, filters);
    });
    if (node)
      return {
        menus: node.data.map((element) => state.menus.details[element]),
        total: node.total,
        loading: state.menus.loading,
      };
    return { menus: [], total: 0, loading: state.menus.loading };
  });
  React.useEffect(() => {
    fetchMenus();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [filters]);

  const fetchMenus = () => {
    dispatch(getMenus(filters));
  };
  return loading ? (
    <Loader />
  ) : (
    <Space direction="vertical">
      <Helmet title={'Menu'} />
      <Row justify="end">
        <Link to="/website/menus/create">
          <Button
            disabled={!(actions.includes('admin') || actions.includes('create'))}
            type="primary"
          >
            New Menu
          </Button>
        </Link>
      </Row>

      <MenuList
        actions={actions}
        data={{ menus, total, loading }}
        filters={filters}
        setFilters={setFilters}
        fetchMenus={fetchMenus}
      />
    </Space>
  );
}

export default Menu;
