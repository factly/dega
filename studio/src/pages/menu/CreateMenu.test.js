import React from 'react';
import { useHistory } from 'react-router-dom';
import { useDispatch, Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import { act } from '@testing-library/react';

import '../../matchMedia.mock';
import CreateMenu from './CreateMenu';
import * as actions from '../../actions/menu';
import MenuCreateForm from './components/MenuForm';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));
jest.mock('react-monaco-editor', () => {
  const MonacoEditor = () => <div />;
  return MonacoEditor;
});
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(),
}));
jest.mock('../../actions/menu', () => ({
  createMenu: jest.fn(),
}));

describe('Menus create component', () => {
  let store;
  let mockedDispatch;
  store = mockStore({
    menus: [],
    details: {},
    loading: false,
  });
  store.dispatch = jest.fn(() => ({}));
  mockedDispatch = jest.fn(() => Promise.resolve({}));
  useDispatch.mockReturnValue(mockedDispatch);

  describe('snapshot testing', () => {
    it('should render the componet', () => {
      const tree = mount(
        <Provider store={store}>
          <CreateMenu />
        </Provider>,
      );
      expect(tree).toMatchSnapshot();
    });
  });
  describe('component testing', () => {
    let wrapper;
    afterEach(() => {
      wrapper.unmount();
    });
    it('should call createMenu', (done) => {
      actions.createMenu.mockReset();
      const push = jest.fn();
      useHistory.mockReturnValueOnce({ push });
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <CreateMenu />
          </Provider>,
        );
      });
      wrapper.find(MenuCreateForm).props().onCreate({ test: 'test' });
      setTimeout(() => {
        expect(actions.createMenu).toHaveBeenCalledWith({ test: 'test' });
        expect(push).toHaveBeenCalledWith('/website/menus');
        done();
      }, 0);
    });
  });
});
