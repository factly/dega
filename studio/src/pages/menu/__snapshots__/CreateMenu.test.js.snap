// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menus create component snapshot testing should render the componet 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <CreateMenu>
    <div>
      <HelmetWrapper
        defer={true}
        encodeSpecialCharacters={true}
        title="Create Menu"
      >
        <SideEffect(NullComponent)
          defer={true}
          encodeSpecialCharacters={true}
          title="Create Menu"
        >
          <NullComponent
            defer={true}
            encodeSpecialCharacters={true}
            title="Create Menu"
          />
        </SideEffect(NullComponent)>
      </HelmetWrapper>
      <MenuForm
        onCreate={[Function]}
      >
        <div>
          <ForwardRef(InternalForm)
            form={
              Object {
                "__INTERNAL__": Object {
                  "itemRef": [Function],
                  "name": undefined,
                },
                "getFieldError": [Function],
                "getFieldInstance": [Function],
                "getFieldValue": [Function],
                "getFieldsError": [Function],
                "getFieldsValue": [Function],
                "getInternalHooks": [Function],
                "isFieldTouched": [Function],
                "isFieldValidating": [Function],
                "isFieldsTouched": [Function],
                "isFieldsValidating": [Function],
                "resetFields": [Function],
                "scrollToField": [Function],
                "setFields": [Function],
                "setFieldsValue": [Function],
                "submit": [Function],
                "validateFields": [Function],
              }
            }
            initialValues={Object {}}
            onFinish={[Function]}
            onValuesChange={[Function]}
          >
            <SizeContextProvider>
              <ForwardRef(Form)
                className="ant-form ant-form-horizontal"
                form={
                  Object {
                    "__INTERNAL__": Object {
                      "itemRef": [Function],
                      "name": undefined,
                    },
                    "getFieldError": [Function],
                    "getFieldInstance": [Function],
                    "getFieldValue": [Function],
                    "getFieldsError": [Function],
                    "getFieldsValue": [Function],
                    "getInternalHooks": [Function],
                    "isFieldTouched": [Function],
                    "isFieldValidating": [Function],
                    "isFieldsTouched": [Function],
                    "isFieldsValidating": [Function],
                    "resetFields": [Function],
                    "scrollToField": [Function],
                    "setFields": [Function],
                    "setFieldsValue": [Function],
                    "submit": [Function],
                    "validateFields": [Function],
                  }
                }
                initialValues={Object {}}
                onFinish={[Function]}
                onFinishFailed={[Function]}
                onValuesChange={[Function]}
              >
                <form
                  className="ant-form ant-form-horizontal"
                  onReset={[Function]}
                  onSubmit={[Function]}
                >
                  <FormItem
                    label="Name"
                    name="name"
                    rules={
                      Array [
                        Object {
                          "message": "Please input the name!",
                          "required": true,
                        },
                      ]
                    }
                  >
                    <WrapperField
                      label="Name"
                      messageVariables={
                        Object {
                          "label": "Name",
                        }
                      }
                      name="name"
                      onReset={[Function]}
                      rules={
                        Array [
                          Object {
                            "message": "Please input the name!",
                            "required": true,
                          },
                        ]
                      }
                      trigger="onChange"
                      validateTrigger="onChange"
                    >
                      <Field
                        fieldContext={
                          Object {
                            "__INTERNAL__": Object {
                              "itemRef": [Function],
                              "name": undefined,
                            },
                            "getFieldError": [Function],
                            "getFieldInstance": [Function],
                            "getFieldValue": [Function],
                            "getFieldsError": [Function],
                            "getFieldsValue": [Function],
                            "getInternalHooks": [Function],
                            "isFieldTouched": [Function],
                            "isFieldValidating": [Function],
                            "isFieldsTouched": [Function],
                            "isFieldsValidating": [Function],
                            "resetFields": [Function],
                            "scrollToField": [Function],
                            "setFields": [Function],
                            "setFieldsValue": [Function],
                            "submit": [Function],
                            "validateFields": [Function],
                            "validateTrigger": "onChange",
                          }
                        }
                        key="_name"
                        label="Name"
                        messageVariables={
                          Object {
                            "label": "Name",
                          }
                        }
                        name={
                          Array [
                            "name",
                          ]
                        }
                        onReset={[Function]}
                        rules={
                          Array [
                            Object {
                              "message": "Please input the name!",
                              "required": true,
                            },
                          ]
                        }
                        trigger="onChange"
                        validateTrigger="onChange"
                        valuePropName="value"
                      >
                        <Row
                          className="ant-form-item"
                          key="row"
                        >
                          <div
                            className="ant-row ant-form-item"
                            style={Object {}}
                          >
                            <FormItemLabel
                              htmlFor="name"
                              label="Name"
                              name="name"
                              prefixCls="ant-form"
                              required={true}
                              requiredMark={true}
                              rules={
                                Array [
                                  Object {
                                    "message": "Please input the name!",
                                    "required": true,
                                  },
                                ]
                              }
                            >
                              <Col
                                className="ant-form-item-label"
                              >
                                <div
                                  className="ant-col ant-form-item-label"
                                  style={Object {}}
                                >
                                  <label
                                    className="ant-form-item-required"
                                    htmlFor="name"
                                    title="Name"
                                  >
                                    Name
                                  </label>
                                </div>
                              </Col>
                            </FormItemLabel>
                            <FormItemInput
                              errors={Array []}
                              label="Name"
                              name={
                                Array [
                                  "name",
                                ]
                              }
                              onDomErrorVisibleChange={[Function]}
                              prefixCls="ant-form"
                              rules={
                                Array [
                                  Object {
                                    "message": "Please input the name!",
                                    "required": true,
                                  },
                                ]
                              }
                              status=""
                              touched={false}
                              validateStatus=""
                              validating={false}
                            >
                              <Col
                                className="ant-form-item-control"
                              >
                                <div
                                  className="ant-col ant-form-item-control"
                                  style={Object {}}
                                >
                                  <div
                                    className="ant-form-item-control-input"
                                  >
                                    <div
                                      className="ant-form-item-control-input-content"
                                    >
                                      <Component
                                        update={1}
                                      >
                                        <Input
                                          id="name"
                                          onChange={[Function]}
                                          placeholder="Enter name"
                                          type="text"
                                        >
                                          <ClearableLabeledInput
                                            bordered={true}
                                            element={
                                              <input
                                                className="ant-input"
                                                id="name"
                                                onBlur={[Function]}
                                                onChange={[Function]}
                                                onFocus={[Function]}
                                                onKeyDown={[Function]}
                                                placeholder="Enter name"
                                                type="text"
                                              />
                                            }
                                            focused={false}
                                            handleReset={[Function]}
                                            id="name"
                                            inputType="input"
                                            onChange={[Function]}
                                            placeholder="Enter name"
                                            prefixCls="ant-input"
                                            triggerFocus={[Function]}
                                            type="text"
                                            value=""
                                          >
                                            <input
                                              className="ant-input"
                                              id="name"
                                              onBlur={[Function]}
                                              onChange={[Function]}
                                              onFocus={[Function]}
                                              onKeyDown={[Function]}
                                              placeholder="Enter name"
                                              type="text"
                                              value=""
                                            />
                                          </ClearableLabeledInput>
                                        </Input>
                                      </Component>
                                    </div>
                                  </div>
                                  <ErrorList
                                    errors={Array []}
                                    onDomErrorVisibleChange={[Function]}
                                  >
                                    <CSSMotion
                                      motionAppear={true}
                                      motionDeadline={500}
                                      motionName="ant-show-help"
                                      onLeaveEnd={[Function]}
                                      removeOnLeave={true}
                                      visible={false}
                                    >
                                      <DomWrapper />
                                    </CSSMotion>
                                  </ErrorList>
                                </div>
                              </Col>
                            </FormItemInput>
                          </div>
                        </Row>
                      </Field>
                    </WrapperField>
                  </FormItem>
                  <FormList
                    name="menu"
                  >
                    <List
                      name="menu"
                    >
                      <WrapperField
                        isList={true}
                        name={Array []}
                        shouldUpdate={[Function]}
                      >
                        <Field
                          fieldContext={
                            Object {
                              "__INTERNAL__": Object {
                                "itemRef": [Function],
                                "name": undefined,
                              },
                              "getFieldError": [Function],
                              "getFieldInstance": [Function],
                              "getFieldValue": [Function],
                              "getFieldsError": [Function],
                              "getFieldsValue": [Function],
                              "getInternalHooks": [Function],
                              "isFieldTouched": [Function],
                              "isFieldValidating": [Function],
                              "isFieldsTouched": [Function],
                              "isFieldsValidating": [Function],
                              "prefixName": Array [
                                "menu",
                              ],
                              "resetFields": [Function],
                              "scrollToField": [Function],
                              "setFields": [Function],
                              "setFieldsValue": [Function],
                              "submit": [Function],
                              "validateFields": [Function],
                              "validateTrigger": "onChange",
                            }
                          }
                          isList={true}
                          key="_"
                          name={Array []}
                          shouldUpdate={[Function]}
                          trigger="onChange"
                          valuePropName="value"
                        >
                          <div>
                            <Button
                              onClick={[Function]}
                            >
                              <Wave>
                                <button
                                  className="ant-btn"
                                  onClick={[Function]}
                                  type="button"
                                >
                                  <LoadingIcon
                                    existIcon={false}
                                    loading={false}
                                    prefixCls="ant-btn"
                                  >
                                    <CSSMotion
                                      motionName="ant-btn-loading-icon-motion"
                                      onAppearActive={[Function]}
                                      onAppearStart={[Function]}
                                      onEnterActive={[Function]}
                                      onEnterStart={[Function]}
                                      onLeaveActive={[Function]}
                                      onLeaveStart={[Function]}
                                      removeOnLeave={true}
                                      visible={false}
                                    >
                                      <DomWrapper />
                                    </CSSMotion>
                                  </LoadingIcon>
                                  <ForwardRef(PlusOutlined)
                                    key=".0"
                                  >
                                    <AntdIcon
                                      icon={
                                        Object {
                                          "icon": Object {
                                            "attrs": Object {
                                              "focusable": "false",
                                              "viewBox": "64 64 896 896",
                                            },
                                            "children": Array [
                                              Object {
                                                "attrs": Object {},
                                                "children": Array [
                                                  Object {
                                                    "attrs": Object {},
                                                    "tag": "style",
                                                  },
                                                ],
                                                "tag": "defs",
                                              },
                                              Object {
                                                "attrs": Object {
                                                  "d": "M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z",
                                                },
                                                "tag": "path",
                                              },
                                              Object {
                                                "attrs": Object {
                                                  "d": "M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z",
                                                },
                                                "tag": "path",
                                              },
                                            ],
                                            "tag": "svg",
                                          },
                                          "name": "plus",
                                          "theme": "outlined",
                                        }
                                      }
                                    >
                                      <span
                                        aria-label="plus"
                                        className="anticon anticon-plus"
                                        role="img"
                                      >
                                        <IconReact
                                          icon={
                                            Object {
                                              "icon": Object {
                                                "attrs": Object {
                                                  "focusable": "false",
                                                  "viewBox": "64 64 896 896",
                                                },
                                                "children": Array [
                                                  Object {
                                                    "attrs": Object {},
                                                    "children": Array [
                                                      Object {
                                                        "attrs": Object {},
                                                        "tag": "style",
                                                      },
                                                    ],
                                                    "tag": "defs",
                                                  },
                                                  Object {
                                                    "attrs": Object {
                                                      "d": "M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z",
                                                    },
                                                    "tag": "path",
                                                  },
                                                  Object {
                                                    "attrs": Object {
                                                      "d": "M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z",
                                                    },
                                                    "tag": "path",
                                                  },
                                                ],
                                                "tag": "svg",
                                              },
                                              "name": "plus",
                                              "theme": "outlined",
                                            }
                                          }
                                        >
                                          <svg
                                            aria-hidden="true"
                                            data-icon="plus"
                                            fill="currentColor"
                                            focusable="false"
                                            height="1em"
                                            key="svg-plus"
                                            viewBox="64 64 896 896"
                                            width="1em"
                                          >
                                            <defs
                                              key="svg-plus-svg-0"
                                            >
                                              <style
                                                key="svg-plus-svg-0-defs-0"
                                              />
                                            </defs>
                                            <path
                                              d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
                                              key="svg-plus-svg-1"
                                            />
                                            <path
                                              d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
                                              key="svg-plus-svg-2"
                                            />
                                          </svg>
                                        </IconReact>
                                      </span>
                                    </AntdIcon>
                                  </ForwardRef(PlusOutlined)>
                                  <span
                                    key=".1"
                                  >
                                     Add menu
                                  </span>
                                </button>
                              </Wave>
                            </Button>
                          </div>
                        </Field>
                      </WrapperField>
                    </List>
                  </FormList>
                  <FormItem
                    label="Metafields"
                    name="meta_fields"
                    style={
                      Object {
                        "marginTop": "20px",
                      }
                    }
                  >
                    <WrapperField
                      label="Metafields"
                      messageVariables={
                        Object {
                          "label": "Metafields",
                        }
                      }
                      name="meta_fields"
                      onReset={[Function]}
                      style={
                        Object {
                          "marginTop": "20px",
                        }
                      }
                      trigger="onChange"
                      validateTrigger="onChange"
                    >
                      <Field
                        fieldContext={
                          Object {
                            "__INTERNAL__": Object {
                              "itemRef": [Function],
                              "name": undefined,
                            },
                            "getFieldError": [Function],
                            "getFieldInstance": [Function],
                            "getFieldValue": [Function],
                            "getFieldsError": [Function],
                            "getFieldsValue": [Function],
                            "getInternalHooks": [Function],
                            "isFieldTouched": [Function],
                            "isFieldValidating": [Function],
                            "isFieldsTouched": [Function],
                            "isFieldsValidating": [Function],
                            "resetFields": [Function],
                            "scrollToField": [Function],
                            "setFields": [Function],
                            "setFieldsValue": [Function],
                            "submit": [Function],
                            "validateFields": [Function],
                            "validateTrigger": "onChange",
                          }
                        }
                        key="_meta_fields"
                        label="Metafields"
                        messageVariables={
                          Object {
                            "label": "Metafields",
                          }
                        }
                        name={
                          Array [
                            "meta_fields",
                          ]
                        }
                        onReset={[Function]}
                        style={
                          Object {
                            "marginTop": "20px",
                          }
                        }
                        trigger="onChange"
                        validateTrigger="onChange"
                        valuePropName="value"
                      >
                        <Row
                          className="ant-form-item"
                          key="row"
                          style={
                            Object {
                              "marginTop": "20px",
                            }
                          }
                        >
                          <div
                            className="ant-row ant-form-item"
                            style={
                              Object {
                                "marginTop": "20px",
                              }
                            }
                          >
                            <FormItemLabel
                              htmlFor="meta_fields"
                              label="Metafields"
                              name="meta_fields"
                              prefixCls="ant-form"
                              required={false}
                              requiredMark={true}
                              style={
                                Object {
                                  "marginTop": "20px",
                                }
                              }
                            >
                              <Col
                                className="ant-form-item-label"
                              >
                                <div
                                  className="ant-col ant-form-item-label"
                                  style={Object {}}
                                >
                                  <label
                                    className=""
                                    htmlFor="meta_fields"
                                    title="Metafields"
                                  >
                                    Metafields
                                  </label>
                                </div>
                              </Col>
                            </FormItemLabel>
                            <FormItemInput
                              errors={Array []}
                              label="Metafields"
                              name={
                                Array [
                                  "meta_fields",
                                ]
                              }
                              onDomErrorVisibleChange={[Function]}
                              prefixCls="ant-form"
                              status=""
                              style={
                                Object {
                                  "marginTop": "20px",
                                }
                              }
                              touched={false}
                              validateStatus=""
                              validating={false}
                            >
                              <Col
                                className="ant-form-item-control"
                              >
                                <div
                                  className="ant-col ant-form-item-control"
                                  style={Object {}}
                                >
                                  <div
                                    className="ant-form-item-control-input"
                                  >
                                    <div
                                      className="ant-form-item-control-input-content"
                                    >
                                      <Component
                                        update={1}
                                      >
                                        <MONACOEditor
                                          id="meta_fields"
                                          language="json"
                                          onChange={[Function]}
                                        >
                                          <MonacoEditor
                                            height="240"
                                            language="json"
                                            onChange={[Function]}
                                            options={
                                              Object {
                                                "autoClosingBrackets": "auto",
                                                "autoClosingQuotes": "auto",
                                                "cursorBlinking": "smooth",
                                                "folding": true,
                                                "lineNumbersMinChars": 4,
                                                "minimap": Object {
                                                  "enabled": false,
                                                },
                                                "scrollBeyondLastLine": false,
                                                "scrollbar": Object {
                                                  "alwaysConsumeMouseWheel": false,
                                                },
                                                "theme": "vs-dark",
                                                "wordWrap": "on",
                                              }
                                            }
                                            width={440}
                                          >
                                            <div />
                                          </MonacoEditor>
                                        </MONACOEditor>
                                      </Component>
                                    </div>
                                  </div>
                                  <ErrorList
                                    errors={Array []}
                                    onDomErrorVisibleChange={[Function]}
                                  >
                                    <CSSMotion
                                      motionAppear={true}
                                      motionDeadline={500}
                                      motionName="ant-show-help"
                                      onLeaveEnd={[Function]}
                                      removeOnLeave={true}
                                      visible={false}
                                    >
                                      <DomWrapper />
                                    </CSSMotion>
                                  </ErrorList>
                                </div>
                              </Col>
                            </FormItemInput>
                          </div>
                        </Row>
                      </Field>
                    </WrapperField>
                  </FormItem>
                  <FormItem>
                    <Row
                      className="ant-form-item"
                      key="row"
                    >
                      <div
                        className="ant-row ant-form-item"
                        style={Object {}}
                      >
                        <FormItemLabel
                          prefixCls="ant-form"
                          requiredMark={true}
                        />
                        <FormItemInput
                          errors={Array []}
                          onDomErrorVisibleChange={[Function]}
                          prefixCls="ant-form"
                          status=""
                          validateStatus=""
                        >
                          <Col
                            className="ant-form-item-control"
                          >
                            <div
                              className="ant-col ant-form-item-control"
                              style={Object {}}
                            >
                              <div
                                className="ant-form-item-control-input"
                              >
                                <div
                                  className="ant-form-item-control-input-content"
                                >
                                  <Button
                                    disabled={true}
                                    htmlType="submit"
                                  >
                                    <Wave>
                                      <button
                                        className="ant-btn"
                                        disabled={true}
                                        onClick={[Function]}
                                        type="submit"
                                      >
                                        <LoadingIcon
                                          existIcon={false}
                                          loading={false}
                                          prefixCls="ant-btn"
                                        >
                                          <CSSMotion
                                            motionName="ant-btn-loading-icon-motion"
                                            onAppearActive={[Function]}
                                            onAppearStart={[Function]}
                                            onEnterActive={[Function]}
                                            onEnterStart={[Function]}
                                            onLeaveActive={[Function]}
                                            onLeaveStart={[Function]}
                                            removeOnLeave={true}
                                            visible={false}
                                          >
                                            <DomWrapper />
                                          </CSSMotion>
                                        </LoadingIcon>
                                        <span
                                          key=".0"
                                        >
                                          Submit
                                        </span>
                                      </button>
                                    </Wave>
                                  </Button>
                                </div>
                              </div>
                              <ErrorList
                                errors={Array []}
                                onDomErrorVisibleChange={[Function]}
                              >
                                <CSSMotion
                                  motionAppear={true}
                                  motionDeadline={500}
                                  motionName="ant-show-help"
                                  onLeaveEnd={[Function]}
                                  removeOnLeave={true}
                                  visible={false}
                                >
                                  <DomWrapper />
                                </CSSMotion>
                              </ErrorList>
                            </div>
                          </Col>
                        </FormItemInput>
                      </div>
                    </Row>
                  </FormItem>
                </form>
              </ForwardRef(Form)>
            </SizeContextProvider>
          </ForwardRef(InternalForm)>
        </div>
      </MenuForm>
    </div>
  </CreateMenu>
</Provider>
`;
